django_oidc_id_provider/views.py

```diff
@@ -0,0 +1,157 @@
+    fc_secret = settings.FC_AS_FS_SECRET
+    state = request.GET.get("state")
+
+    try:
+        connection = Connection.objects.get(state=state)
+    except Connection.DoesNotExist:
+        log.info("FC as FS - This state does not seem to exist")
+        log.info(state)
+        return HttpResponseForbidden()
+
+    if connection.is_expired:
+        log.info("408: FC connection has expired.")
+        return render(request, "408.html", status=408)
+
+    code = request.GET.get("code")
+    if not code:
+        log.info("403: No code has been provided.")
+        return HttpResponseForbidden()
+
+    token_url = f"{fc_base}/token"
+    payload = {
+        "grant_type": "authorization_code",
+        "redirect_uri": fc_callback_uri,
+        "client_id": fc_id,
+        "client_secret": fc_secret,
+        "code": code,
+    }
+    headers = {"Accept": "application/json"}
+
+    request_for_token = python_request.post(token_url, data=payload, headers=headers)
+
+    def fc_error(log_msg):
+        log.error(log_msg)
+        django_messages.error(
+            request,
+            "Nous avons rencontré une erreur en tentant d'interagir avec "
+            "France Connect. C'est probabablement temporaire. Pouvez-vous réessayer "
+            "votre requête ?",
+        )
+
+        return redirect(reverse("new_mandat"))
+
+    try:
+        content = request_for_token.json()
+    except ValueError:  # not a valid JSON
+        return fc_error(
+            f"Request to {token_url} failed. Status code: "
+            f"{request_for_token.status_code}, body: {request_for_token.text}"
+        )
+
+    connection.access_token = content.get("access_token")
+    if connection.access_token is None:
+        return fc_error(
+            f"No access_token return when requesting {token_url}. JSON response: "
+            f"{repr(content)}"
+        )
+
+    connection.save()
+    fc_id_token = content.get("id_token")
+
+    try:
+        decoded_token = jwt.decode(
+            fc_id_token,
+            settings.FC_AS_FS_SECRET,
+            audience=settings.FC_AS_FS_ID,
+            algorithm="HS256",
+        )
+    except ExpiredSignatureError:
+        log.info("403: token signature has expired.")
+        return HttpResponseForbidden()
+
+    if connection.nonce != decoded_token.get("nonce"):
+        log.info("403: The nonce is different than the one expected.")
+        return HttpResponseForbidden()
+
+    if connection.is_expired:
+        log.info("408: FC connection has expired.")
+        return render(request, "408.html", status=408)
+
+    logout_base = f"{fc_base}/logout"
+    logout_id_token = f"id_token_hint={fc_id_token}"
+    logout_state = f"state={state}"
+    logout_redirect = f"post_logout_redirect_uri={fc_callback_uri_logout}"
+    logout_url = f"{logout_base}?{logout_id_token}&{logout_state}&{logout_redirect}"
+    return redirect(logout_url)
+
+
+def get_user_info(connection: Connection) -> tuple:
+    return ()
```

Comment:
@mrjmad @tut-tuuut @Anna-Livia Ici, les vues qui a été écrite pour Aidants Connect font plein d'opérations très spécifiques à Aidant Connect. Par exemple [ça](https://github.com/betagouv/Aidants_Connect/blob/main/aidants_connect_web/views/FC_as_FS.py#L107) ou [ça](https://github.com/betagouv/Aidants_Connect/blob/main/aidants_connect_web/views/FC_as_FS.py#L151-L157). Pour l'instant, l'idée que j'ai en tête, pour permettre, de généraliser ça, c'est de passer sur des class-based-views et permettre de configurer la classe de la vue avec un paramètre dans le `settings.py`. Il suffira de déplacer les trucs un peu spécifiques dans des fonctions « callback » de la CBV. Si vous avez une meilleure idée, dites-moi.

Reply:
J'avoue que le seul mot que j'ai compris c'est "callback".
Et émettre des signaux que les classes métier pourraient attraper, ça se ferait ?

Reply:
Je sais pas trop si les signaux sont synchrônes. Dans le doute, je dois considérer que non. Et je crois que dans le code, y'a des endroits où je vais avoir besoin d'un résultat pour continuer l'exécution :thinking:

Reply:
Une rapide recherche semble dire que les signaux sont bien synchrones, sauf si on se complique beaucoup la vie pour qu'ils ne le soient pas. Mais en effet si tu as besoin du résultat ça n'est pas la solution.
Sinon j'aime bien la manière dont c'est fait dans django-import-export, et en effet ça impliquerait d'utiliser des classes à la place des fonctions habituelles. Après c'est plutôt bien documenté dans la doc officielle, donc why not ?

Reply:
Après avoir réfléchi : 
- changer toute la CBV à travers le settings, ne me plaît qu'à moitié. Je m'explique. la CBV va être le cœur du truc. c'est elle qui gère tout et c'est la classe que l'on va demander aux gens de modifier. 
- pour les signaux, on peut récupérer le résultat oui ( voir ici : https://docs.djangoproject.com/en/3.2/topics/signals/#sending-signals) . La question est de savoir si on peut passer toutes les informations que l'on veut au signal ou pas. Et de savoir si conceptuellement, ça nous va d'avoir plein de petites fonctions pour faire plein de petite modification. 
- Pour l'utilisation de CBV de toute façon c'est "normalement" de puis très très longtemps "la bonne pratique", donc mettre un peu des CBV c'est toujours une bonne chose :) [ bon d'accord j'aime bien les CBV ]
- j'aime bien aussi la méthode de django-import-export. Tel que je le comprends dans une optique ici on aurait une CBV qui fait tout le taff "métier normal de la gestion du truc" mais qui a un moment va récupérer un objet "métier supplémentaire" ( en fait un truc genre get_addition_workflow_klass(). et qui au bon moment ferrait un get_addition_workflow_klass().after_connexion() ou get_addition_workflow_klass().after_received_callback() etc etc.. 

du coup le get_addition_workflow_klass pourrait aller prendre une valeur par defaut DefaultAdditionnalWorkflowClass() qui définierait toutes les méthodes de truc additionnel en pass. ou genre settings.AdditionalWorkFlowClass

et du coup pour étendre le fonctionnement nonimal il "suffit" de définir une classe MySpecificAdditionalWorkflow(DefaultAdditionnalWorkflowClass) et d'écrire que les méthodes qui nous interressent et de poser cette classe en settings. 

Vous en pensez quoi ?

Reply:
Donc, si je comprends bien `django-import-export` n'utilise pas de paramètre dans `settings.py` pour découvrir la classe qui contient les callbacks, elle est auto-découverte ? Possiblement avec `__subclasses__`, j'imagine ? Je vais aller voir le code alors.

Reply:
En fait dans django-import-export on étend dans notre code les classes du module, qui contiennent un certain nombre de méthodes vides. On redéfinit juste les méthodes qui nous intéressent. C'est de l'héritage tout simple, ya pas de `__subclasses__` ou de paramétrage.

Reply:
Oui mais `django-import-export` doit découvrir les classes dans notre code avec `__subclasses__`, justement. Puisqu'il n'y a pas de paramétrage :thinking: 

En fait, j'irai voir le code, tout simplement :sweat_smile:

Reply:
C'est nous qui découvrons le code de Django-import-export en surchargeant les classes que le module propose, en fait. Lui il ne fait rien tout seul tant que l'on ne lui demande rien. Magicauth fonctionne sur le même principe j'ai l'impression, il n'y a aucune mention de __subclasses__ nulle part !

Reply:
Chuis pas sûr que la manière de faire de `django-import-export` me serve à grand-chose en fait. Tout semble reposer sur le fait que Django découvre et instancie tout ce qui se trouve dans `admin.py`. Je pourrais fonctionner pareil en disant qu'il faut placer les comportements custom dans, mettons, `oidc.py`.